### DISCUSSION TITLE
arm64 M1 mac compilation guidelines

### DISCUSSION THREAD

**Discussion by nomar500** [Beep6581/RawTherapee#6174] [Q&A]
Dear community, 

I've been trying to build a stable version of RT-5.8 dev for my M1 mac.
Much of the struggle I've faced is in #6152 but it's not the ideal thread for asking all my beginner questions. [discuss forum](https://discuss.pixls.us/t/re-test-of-rawtherapee-on-macos-11-1-big-sur/23347/16) is not better, either.

- Can we please work on up-to-date guidelines for M1 mac targets? [RawPedia](http://rawpedia.rawtherapee.com/MacOS#Compile_RawTherapee) needs a bit of work to be adapted for Big Sur and arm64, if I am not mistaken.
- also, since rawpedia is more "visible" to end-users of RT, should we not include details pertaining to compilation choices for Mac OS as well? (`llvm` _VS_ Apple Native CLang, etc. what are the trade-offs?)

I'll edit this question if need be, mostly with content from the aforementioned case.

Thanks for your help.

<details>
<summary>EDIT (unfold me): adding where I am, currently</summary>

##  A. Compile RawTherapee 

FIXME Dependencies: I don't have a clear view on those 😒 due to my history with `llvm`

### A.1. with `Apple CLang`

> valid for Mac OS 11.2, RawTherapee 5.8 dev, arm64 M1 mac architecture
> limitations:
>  - `fftw3f` not supported. Impact: Dynamic Range Compression tool speed is reduced.
>  - `LTO` not supported. Impact: overall speed reduced. 

-  at the root of the RawTherapee repository, you can create a script `make_RT.sh` to create the "build" folder if it does not exist already, purge its contents and compile rawtherapee. You'll then be able to launch RawTherapee from the terminal.

```bash
#!/bin/bash

mkdir -p build 
cd build
pwd
sudo rm -r *
eval $(/opt/homebrew/bin/brew shellenv)

cmake --log-level=VERBOSE --log-context -DCMAKE_BUILD_TYPE="Release" \
      -DLOCAL_PREFIX="$(brew --prefix)" \
      -DOSX_DEV_BUILD="ON" \
      -DPROC_TARGET_NUMBER="2" \
      -DCACHE_NAME_SUFFIX="5.8-dev" \
      -DWITH_LTO="OFF" \
      -DCMAKE_C_COMPILER="clang" \
      -DCMAKE_CXX_COMPILER="clang++" \
      -DLENSFUNDBDIR="/Applications/RawTherapee.app/Contents/Resources/share/lensfun" \
      -DCMAKE_OSX_DEPLOYMENT_TARGET="11.2" \
      -S .. --trace-source=../CMakeLists.txt --trace-expand

make -j$(sysctl -n hw.ncpu) install
#sudo make macosx_bundle
```

- run `make_RT.sh` (logs appear on screen and are written to `make_RT.log`, too.

```bash
bash make_RT.sh |& tee make_RT.log
```

- launch RawTherapee

```bash
build/Release/MacOS/rawtherapee
```

#### A.2.1. why you would want to use `llvm`
TODO

#### A.2.2. `make_RT.sh`
FIXME - this does not work
> valid for Mac OS 11.2, RawTherapee 5.8 dev, arm64 M1 mac architecture
> limitations:
>  - `fftw3f` not supported. Impact: Dynamic Range Compression tool speed is reduced.
>  - `LTO` not supported. Impact: overall speed reduced. 

- create an `make_RT.sh` script

```bash
#!/bin/bash

mkdir -p build 
cd build
pwd
sudo rm -r *

which brew || exit 2

eval $(/opt/homebrew/bin/brew shellenv)
export PATH="$(brew --prefix llvm)/bin:$PATH" && echo "PATH: $PATH" 
export CC="$(brew --prefix llvm)/bin/clang" && echo "CC: $CC"
export CXX="${CC}++" && echo "CXX: $CXX"
export CMAKE_LD="$(brew --prefix llvm)/bin/lld" 
export CMAKE_AR="$(brew --prefix llvm)/bin/llvm-ar" && echo "CMAKE_AR: $CMAKE_AR" 
export CMAKE_RANLIB="$(brew --prefix llvm)/bin/llvm-ranlib" && echo "CMAKE_RANLIB: $CMAKE_RANLIB" 
export PKG_CONFIG_PATH="/opt/homebrew/opt/libffi/lib/pkgconfig:/opt/homebrew/opt/expat/lib/pkgconfig"
export LDFLAGS="-v -L$(brew --prefix llvm)/lib -Wl,-rpath,$(brew --prefix llvm)/lib -L/opt/homebrew/opt/libffi/lib -L/opt/homebrew/opt/expat/lib -L/opt/homebrew/lib" && \
export CPPFLAGS="-v -I$(brew --prefix llvm)/include -I$(brew --prefix llvm)/include/c++/v1/ -I/opt/homebrew/opt/libffi/include -I/opt/homebrew/opt/expat/include" && \
cmake --log-level=VERBOSE -DCMAKE_BUILD_TYPE="Release" \
  -DLOCAL_PREFIX="$(brew --prefix)" \
  -DCMAKE_OSX_DEPLOYMENT_TARGET="11.2" \
  -DOSX_DEV_BUILD="ON" \
  -DPROC_TARGET_NUMBER="2" \
  -DCACHE_NAME_SUFFIX="5.8-dev" \
  -DLENSFUNDBDIR="/Applications/RawTherapee.app/Contents/Resources/share/lensfun" \
  -DCMAKE_C_COMPILER="$CC" \
  -DCMAKE_CXX_COMPILER="$CXX" \
  -DOpenMP_C_LIB_NAMES:STRING="libomp" \
  -DOpenMP_CXX_LIB_NAMES:STRING="libomp" \
  -DWITH_LTO="ON" \
  -DOpenMP_libomp_LIBRARY:FILEPATH="/opt/homebrew/lib/libomp.dylib" \
  -DOpenMP_CXX_FLAGS:STRING="-Wno-pass-failed -Wno-deprecated-register -Xpreprocessor -fopenmp" \
  -DOpenMP_C_FLAGS:STRING="-Wno-pass-failed -Wno-deprecated-register -Xpreprocessor -fopenmp" \
  -DOpenMP_CXX_INCLUDE_DIR:PATH="/opt/homebrew/include" \
  -DCMAKE_VERBOSE_MAKEFILE="FALSE" \
  -DCMAKE_EXE_LINKER_FLAGS="$LDFLAGS" \
  -DCMAKE_CXX_FLAGS="$CPPFLAGS" \
  -DCMAKE_AR="${CMAKE_AR}" \
  -DCMAKE_RANLIB="${CMAKE_RANLIB}" \
  -DCMAKE_LINKER="$(brew --prefix llvm)/bin/lld" \
  -S .. --trace-source=../CMakeLists.txt --trace-expand
make -j$(sysctl -n hw.ncpu) install
#sudo make macosx_bundle
```

</details>

**Comment by Thanatomanic** [comment]
While discussing here is perfectly fine, the best support is still found on on Pixls (larger userbase). Try to see if you can get an experienced macOS builder to talk with you privately if you keep running into trouble. Maybe contact @Benitoite (HIRAM on Pixls)? Or @stwunsch, since he has managed to make successful builds on M1.

Your point about documentation is valid, but can only be handled _after_ we know exactly what is needed for a successful compilation.
Additionally, since you're a macOS user, and many of us are not, it would be most helpful if you actually contribute to the documentation (propose texts). Then again, that's not everbody's cup of tea...

**Reply by nomar500** [reply]
hi @Thanatomanic, 
for support, that's what I thought and that's why I initially asked for help in the PIXLS forum (`@nonophuran`), [here](https://discuss.pixls.us/t/test-of-rawtherapee-on-macos-11-1-big-sur/22389/38), without much success in the end.

I then sifted through the code and the error log files, contacted @Floessie and was pointed to @stwunsch. 

Instead of keeping it private, I've reached #6152 and elaborated on my various issues. I hope I'm making sensible points.
Since my questions address a variety of topics, I thought it would be more appropriate not to flood a PR discussion and to create a separate thread here, before suggesting documentation "fixes".

**Comment by nomar500** [comment]
note: I've just edited the original post, to add what could be included into rawpedia (see the collapsible section)
